drake lint handles a missing terminating newline badly

When invoking the drakelint test on a file that is missing a terminating newline, drake lint gets *really* angry.  Cpplint handles it quite gracefully.

Example output:

```
FAIL: //geometry/proximity:py/collision_filter_legacy_drakelint (see /home/seancurtis/.cache/bazel/_bazel_seancurtis/c482a09839de82c7b8c8a508d26ab2c3/execroot/drake/bazel-out/k8-opt/testlogs/geometry/proximity/py/collision_filter_legacy_drakelint/test.log)
INFO: From Testing //geometry/proximity:py/collision_filter_legacy_drakelint:
==================== Test output for //geometry/proximity:py/collision_filter_legacy_drakelint:
drakelint.py: Linting geometry/proximity/collision_filter_legacy.h
Traceback (most recent call last):
  File "/home/seancurtis/.cache/bazel/_bazel_seancurtis/c482a09839de82c7b8c8a508d26ab2c3/sandbox/linux-sandbox/10/execroot/drake/bazel-out/k8-opt/bin/geometry/proximity/py/collision_filter_legacy_drakelint.runfiles/drake/tools/lint/drakelint.py", line 108, in <module>
    main()
  File "/home/seancurtis/.cache/bazel/_bazel_seancurtis/c482a09839de82c7b8c8a508d26ab2c3/sandbox/linux-sandbox/10/execroot/drake/bazel-out/k8-opt/bin/geometry/proximity/py/collision_filter_legacy_drakelint.runfiles/drake/tools/lint/drakelint.py", line 99, in main
    total_errors += _check_includes(filename)
  File "/home/seancurtis/.cache/bazel/_bazel_seancurtis/c482a09839de82c7b8c8a508d26ab2c3/sandbox/linux-sandbox/10/execroot/drake/bazel-out/k8-opt/bin/geometry/proximity/py/collision_filter_legacy_drakelint.runfiles/drake/tools/lint/drakelint.py", line 40, in _check_includes
    tool = IncludeFormatter(filename)
  File "/home/seancurtis/.cache/bazel/_bazel_seancurtis/c482a09839de82c7b8c8a508d26ab2c3/sandbox/linux-sandbox/10/execroot/drake/bazel-out/k8-opt/bin/geometry/proximity/py/collision_filter_legacy_drakelint.runfiles/drake/tools/lint/formatter.py", line 243, in __init__
    super(IncludeFormatter, self).__init__(filename, *args, **kwargs)
  File "/home/seancurtis/.cache/bazel/_bazel_seancurtis/c482a09839de82c7b8c8a508d26ab2c3/sandbox/linux-sandbox/10/execroot/drake/bazel-out/k8-opt/bin/geometry/proximity/py/collision_filter_legacy_drakelint.runfiles/drake/tools/lint/formatter.py", line 48, in __init__
    self._check_rep()
  File "/home/seancurtis/.cache/bazel/_bazel_seancurtis/c482a09839de82c7b8c8a508d26ab2c3/sandbox/linux-sandbox/10/execroot/drake/bazel-out/k8-opt/bin/geometry/proximity/py/collision_filter_legacy_drakelint.runfiles/drake/tools/lint/formatter.py", line 56, in _check_rep
    assert line.endswith("\n"), line
AssertionError: }  // namespace drake
================================================================================
```

--- a/tools/lint/drakelint.py
+++ b/tools/lint/drakelint.py
@@ -37,7 +37,11 @@ def _check_invalid_line_endings(filename):
 
 def _check_includes(filename):
     """Return 0 if clang-format-includes is a no-op, and 1 otherwise."""
-    tool = IncludeFormatter(filename)
+    try:
+        tool = IncludeFormatter(filename)
+    except Exception as e:
+        print("ERROR: " + filename + ":0: " + e.message)
+        return 1
     tool.format_includes()
     first_difference = tool.get_first_differing_original_index()
     if first_difference is not None:

--- a/tools/lint/formatter.py
+++ b/tools/lint/formatter.py
@@ -45,9 +45,11 @@ class FormatterBase(object):
         else:
             self._original_lines = [unicode(line) for line in readlines]
         self._working_lines = list(self._original_lines)
-        self._check_rep()
         if any(["\r" in line for line in self._working_lines]):
             raise Exception("DOS newlines are not supported")
+        if not self._working_lines[-1].endswith("\n"):
+            raise Exception("Missing newline character at end of file")
+        self._check_rep()
 
     def _check_rep(self):
         assert self._filename

--- a/tools/lint/test/formatter_test.py
+++ b/tools/lint/test/formatter_test.py
@@ -65,6 +65,13 @@ class TestFormatterBase(unittest.TestCase):
         with self.assertRaisesRegexp(Exception, "DOS newline"):
             FormatterBase("filename.cc", readlines=original_lines)
 
+    def test_missing_eof(self):
+        original_lines = [
+            '#include "line0"',
+        ]
+        with self.assertRaisesRegexp(Exception, "newline.*end of file"):
+            FormatterBase("filename.cc", readlines=original_lines)
+
 
 class TestIncludeFormatter(unittest.TestCase):
 

